internal/providers/vault/ca.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"time"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/cfg"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+func certificateAuthorityProvider(ctx context.Context, conf cfg.Configuration, secretId secrets.SecretIdentifier, cfg *vault.CertificateAuthority) ([]byte, error) {
+	vaultConfig, ok := GetVaultConfig(conf)
+	if !ok || vaultConfig == nil {
+		return nil, fnerrors.BadInputError("invalid vault app role configuration: missing provider configuration")
+	}
+
+	vaultClient, err := login(ctx, vaultConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.CommonName == "" {
+		return nil, fnerrors.BadInputError("invalid certificate authority configuration: missing common name")
+	}
+	if cfg.Ttl == "" {
+		return nil, fnerrors.BadInputError("invalid certificate authority configuration: missing ttl")
+	}
+
+	if _, err := time.ParseDuration(cfg.Ttl); err != nil {
+		return nil, fnerrors.BadInputError("ttl is not a valid duration")
```

Comment:
Can you wrap the error, maybe it provides some useful details?

Reply:
done